egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_train.py

```diff
@@ -0,0 +1,628 @@
+                    accum_grad: int,
+                    att_rate: float,
+                    current_epoch: int,
+                    tb_writer: SummaryWriter,
+                    num_epochs: int,
+                    global_batch_idx_train: int,
+                    world_size: int,
+                    rank: int,
+                    ):
+    """One epoch training and validation.
+
+    Args:
+        dataloader: Training dataloader
+        valid_dataloader: Validation dataloader
+        model: Acoustic model to be trained
+        device: Training device, torch.device("cpu") or torch.device("cuda", device_id)
+        numericalizer: Convert transcript from text --> tokens --> token_ids
+        optimizer: Training optimizer
+        accum_grad: Number of gradient accumulation
+        att_rate: Attention loss rate, final loss is att_rate * att_loss + (1-att_rate) * ctc_loss
+        current_epoch: current training epoch, for logging only
+        tb_writer: tensorboard SummaryWriter
+        num_epochs: total number of training epochs, for logging only
+        global_batch_idx_train: global training batch index before this epoch, for logging only
+        world_size: indicate multi-gpu training or not
+        rank: logging.info will works only rank==0 to avoid duplicate information.
+
+    Returns:
+        A tuple of 3 scalar:  (total_objf / epoch_num_utts, valid_average_objf, global_batch_idx_train)
+        - `total_objf / total_frames` is the average training loss
+        - `valid_average_objf` is the average validation loss
+        - `global_batch_idx_train` is the global training batch index after this epoch
+    """
+    # Return value
+    total_objf, epoch_num_utts = 0., 0.
+
+    # To record loginterval batches
+    loginterval = 100
+    loginterval_loss, loginterval_att_loss, loginterval_ctc_loss, loginterval_num_utts = 0., 0., 0., 0.
+
+    valid_average_objf = float('inf')
+    time_waiting_for_batch = 0
+    forward_count = 0
+    prev_timestamp = datetime.now()
+
+    model.train()
+    for batch_idx, batch in enumerate(dataloader):
+        forward_count += 1
+        if forward_count == accum_grad:
+            is_update = True
+            forward_count = 0
+        else:
+            is_update = False
+
+        global_batch_idx_train += 1
+        timestamp = datetime.now()
+        time_waiting_for_batch += (timestamp - prev_timestamp).total_seconds()
+
+        curr_batch_objf, curr_batch_att_loss, curr_batch_ctc_loss, curr_batch_num_utts = get_objf(
+            batch=batch,
+            model=model,
+            device=device,
+            numericalizer=numericalizer,
+            is_training=True,
+            is_update=is_update,
+            accum_grad=accum_grad,
+            att_rate=att_rate,
+            tb_writer=tb_writer,
+            global_batch_idx_train=global_batch_idx_train,
+            optimizer=optimizer,
+        )
+
+        total_objf += curr_batch_objf
```

Comment:
I think this should be: total_objf += curr_batch_objf * curr_batch_num_utts, because you'll later be normalizing by epoch_num_utts.

Reply:
fixed.